pulp_smash/tests/docker/cli/test_copy.py

```diff
@@ -0,0 +1,244 @@
+# coding=utf-8
+"""Tests for copying docker units between repositories."""
+from __future__ import unicode_literals
+import re
+
+import unittest2
+from packaging.version import Version
+
+from pulp_smash import utils
+from pulp_smash.tests.docker.cli import utils as docker_utils
+
+
+class CopyAllImagesTestCase(docker_utils.BaseTestCase,
+                            docker_utils.SuccessMixin):
+    """Test copying all Images from one repository to another."""
+
+    @classmethod
+    def setUpClass(cls):
+        """Create and sync a docker repository with a v1 registry."""
+        super(CopyAllImagesTestCase, cls).setUpClass()
+        if cls.cfg.version < Version('2.8'):
+            raise unittest2.SkipTest('These tests require Pulp 2.8 or above.')
+        # Create and sync a repository with a v1 feed to bring some Images in.
+        docker_utils.create_repo(
+            cls.cfg, cls.repo_id, upstream_name='library/busybox',
+            sync_v1=True, sync_v2=False)
+        docker_utils.sync_repo(cls.cfg, cls.repo_id)
+
+        # Now create a feedless repository and copy the images from the first
+        # repository over.
+        cls.copy_target = utils.uuid4()
+        docker_utils.create_repo(cls.cfg, cls.copy_target)
+        cls.completed_proc = docker_utils.copy(cls.cfg, 'image', cls.repo_id,
+                                               cls.copy_target)
+
+    @classmethod
+    def tearDownClass(cls):
+        """Delete the copy_target repo."""
+        super(CopyAllImagesTestCase, cls).tearDownClass()
+        docker_utils.delete_repo(cls.cfg, cls.copy_target)
+
+    def test_positive_copy_output(self):
```

Comment:
In your cited example, if the tags aren't in dest, the manifests won't be either. We will know there is a problem and we will fix it. I don't think that's a benefit.

Also, the general pattern in pulp-smash of performing all the "work" in setUpClass means that we need a lot more classes. This means that we sync from docker n times instead of one time in this copy module. I'm not a fan of that, and I'm sure Docker wouldn't be either. If anything, we should have one setUpClass that does the sync, and then each test should do whatever copying it needs along with whatever assertions it needs.

It's easy to add assertions to tests, so I don't think that benefit is meaningfully different between the two approaches.